packages/bff/src/services/toolsService.ts

```diff
@@ -0,0 +1,393 @@
+): bffApi.TokenGenerationValidationResult {
+  const clientAssertionErrors = errs.clientAssertionErrors ?? [];
+  const keyRetrieveErrors = errs.keyRetrieveErrors ?? [];
+  const clientSignatureErrors = errs.clientSignatureErrors ?? [];
+  const platformStateErrors = errs.platformStateErrors ?? [];
+
+  return {
+    clientKind,
+    eservice,
+    steps: {
+      clientAssertionValidation: {
+        result: getStepResult([], clientAssertionErrors),
+        failures: apiErrorsToValidationFailures(clientAssertionErrors),
+      },
+      publicKeyRetrieve: {
+        result: getStepResult(clientAssertionErrors, keyRetrieveErrors),
+        failures: apiErrorsToValidationFailures(keyRetrieveErrors),
+      },
+      clientAssertionSignatureVerification: {
+        result: getStepResult(
+          [...clientAssertionErrors, ...keyRetrieveErrors],
+          clientSignatureErrors
+        ),
+        failures: apiErrorsToValidationFailures(clientSignatureErrors),
+      },
+      platformStatesVerification: {
+        result: getStepResult(
+          [
+            ...clientAssertionErrors,
+            ...keyRetrieveErrors,
+            ...clientSignatureErrors,
+          ],
+          platformStateErrors
+        ),
+        failures: apiErrorsToValidationFailures(platformStateErrors),
+      },
+    },
+  };
+}
+
+function getStepResult(
+  prevStepErrors: Array<ApiError<string>>,
+  currentStepErrors: Array<ApiError<string>>
+): bffApi.TokenGenerationValidationStepResult {
+  if (currentStepErrors.length > 0) {
+    return bffApi.TokenGenerationValidationStepResult.Enum.FAILED;
+  } else if (prevStepErrors.length > 0) {
+    return bffApi.TokenGenerationValidationStepResult.Enum.SKIPPED;
+  } else {
+    return bffApi.TokenGenerationValidationStepResult.Enum.PASSED;
+  }
+}
+
+async function retrieveKey(
+  {
+    authorizationClient,
+    purposeProcessClient,
+    agreementProcessClient,
+    catalogProcessClient,
+  }: PagoPAInteropBeClients,
+  jwt: ClientAssertion,
+  ctx: WithLogger<BffAppContext>
+): Promise<
+  SuccessfulValidation<ApiKey | ConsumerKey> | FailedValidation<ErrorCodes>
+> {
+  const client = await authorizationClient.token
+    .getKeyWithClientByKeyId({
+      params: {
+        clientId: jwt.payload.sub,
+        keyId: jwt.header.kid,
+      },
+      headers: ctx.headers,
+    })
+    .catch(() => undefined);
+
+  if (!client) {
+    return {
+      data: undefined,
+      errors: [
+        clientAssertionPublicKeyNotFound(jwt.header.kid, jwt.payload.sub),
+      ],
+    };
+  }
+
+  const { encodedPem } = await authorizationClient.client.getClientKeyById({
+    headers: ctx.headers,
+    params: {
+      clientId: client.client.id,
+      keyId: jwt.header.kid,
+    },
+  });
+
+  if (!jwt.payload.purposeId) {
+    return {
+      data: undefined,
+      errors: [purposeIdNotFoundInClientAssertion()],
+    };
+  }
+  const purposeId = unsafeBrandId<PurposeId>(jwt.payload.purposeId);
```

Comment:
Could this check be moved at the top, so we avoid the previous API calls in case there is no purposeId?